service-base/src/main/java/org/eclipse/hono/service/http/HttpUtils.java

```diff
@@ -181,7 +181,7 @@ public static void failWithHeaders(
         if (ctx.failed() || ctx.response().ended()) {
             // nothing to do
         } else {
-            if (headers != null) {
+            if (headers != null && !ctx.response().closed()) {
```

Comment:
shouldn't this better be part of the condition in the first if statement?

Reply:
No, if the response is closed but not failed/ended yet, the `fail` method should get called here (which then also causes the `TracingHandler` failureHandler to get called).

The extra `closed` check above is to ensure that the `putHeader` invocation doesn't throw an exception. It would do so for a closed response using Vert.x 3.7.1 (Hono 1.0). In Vert.x 3.8.4 (Hono master) this check isn't strictly needed any more, but I think the extra condition doesn't hurt here.

Reply:
I see, then adding a corresponding comment would improve readability a lot here FMPOV.

Reply:
Yes, I've added comments.